content/blog/2021-07-06-hyperparam-tuning.md

```diff
@@ -0,0 +1,241 @@
+    subprocess.run(["dvc", "exp", "run", "--queue", "--set-param", f"train.n_est={val}"])
+
+for val in min_split_values:
+    subprocess.run(["dvc", "exp", "run", "--queue", "--set-param", f"train.min_split={val}"])
+```
+
+This is a simple grid search. We have two hyperparameters we want to tune:
+`n_est` and `min_split`. So we have arrays with a few values in them to mimic
+the exhaustive search a grid search can handle. Then we loop through the values
+and create queued experiments for them using `subprocess`.
+
+You can run this script now and generate your queue with this command.
+
+`python src/tuning.py`
+
+You'll see some outputs in the terminal telling you that your experiments have
+been queued. Then you can run them all with the following command.
+
+`dvc exp run --run-all`
+
+This will run every experiment that has been queued. Once all of those have run,
+take a look at your metrics for each experiment.
+
+`dvc exp show`
+
+Your table should look similar to this when you run `dvc exp show`.
+
+```bash
+table
+```
+
+Now you can see how your precision changed with each hyperparameter value
+update. This is a quick implementation of grid search in DVC. You could read the
+hyperparameter values from a different file or data source or make this tuning
+script as fancy as you like. The main thing you need is the
+`dvc exp run --queue -set-param <param>` command to execute when you add new
+values.
+
+### Random Search
+
+Another commonly used method for tuning hyperparameters is random search. This
+takes random values for hyperparameters and builds the model with them. It's not
+an exhaustive search like grid search and it typically outperforms grid search,
+especially when a small subset of the hyperparameters produce the biggest
+difference in your model.
+
+We're going to add an example of random search to the `tuning.py` file we
+created for grid search. This will add queued experiments with the randomly
+selected hyperparameter values. Add the following code to `tuning.py` below the
+grid search implementation.
+
+```python
+# Automated random search experiments
+rand_n_est_values = random.sample(range(250, 500), 10)
+rand_min_split_values = random.sample(range(100, 150), 10)
+
+for val in rand_n_est_values:
+    subprocess.run(["dvc", "exp", "run", "--queue", "--set-param", f"train.n_est={val}"])
+
+for val in rand_min_split_values:
+    subprocess.run(["dvc", "exp", "run", "--queue", "--set-param", f"train.min_split={val}"])
+```
+
+This random search could be far more complex with a Gaussian or Bayesian
+distribution to handle the hyperparameter value selections, but we're keeping it
+super simple by choosing random numbers to focus on reproducibility. This will
+generate ten experiments with random values for each hyperparameter.
+
+You can run these new experiments with `dvc exp run --run-all` and then take a
+look at the results with `dvc exp show`. Your table should look something like
+this.
+
+```bash
+table
+```
+
```

Comment:
would be nice to have a 3D plot here (e.g. `avg_prec` vs `n_est` & `min_split`)

Reply:
You could do a parallel coordinates plot here (scroll down to the bottom of https://colab.research.google.com/drive/1_Unynd2P2tDtJDhB5Iql-H7Nk5YLCRpF#scrollTo=Yq-5oIs0nsHX), but it's probably out of scope for this post, since it will likely prompt questions about how to generate it.

Reply:
I'll get that table added here, but I don't want to introduce something else that takes them away from DVC. Maybe we could do a different post on showing the metrics using DVC experiments?